src/components/Restore.tsx

```diff
@@ -5,38 +5,54 @@ import { SimpleTextInput } from './SimpleTextInput';
 import { Debug } from '../Debug';
 import { Colors, DefaultNavigationBarHeight } from '../styles';
 import { Button } from './Button';
-import { isValidBackup, restoreBackupToString } from '../BackupRestore';
-import { stringToHex } from '../conversion';
+import { restoreBinaryBackupToString } from '../BackupRestore';
 import { getAppStateFromSerialized } from '../reducers';
 import { TypedNavigation } from '../helpers/navigation';
+import * as Swarm from '../swarm/Swarm';
+import { AppState } from '../reducers/AppState';
+import { HexString } from '../opaqueTypes';
+import { decrypt, ENCRYPTED_HEX_HEADER_LENGTH } from '../cryptoHelpers';
+import { stringToByteArray, hexToByteArray, byteArrayToHex } from '../conversion';
+
+const BACKUP_DATA_LENGTH = 128;
+const ENCRYPTED_BACKUP_DATA_LENGHT = ENCRYPTED_HEX_HEADER_LENGTH + BACKUP_DATA_LENGTH;
+const CONTENT_HASH_OFFSET = 0;
+const SECRET_OFFSET = 64;
 
 export interface StateProps {
     navigation: TypedNavigation;
+    swarmGatewayAddress: string;
 }
 
 export interface DispatchProps {
-    onRestoreData: (data: string, secretHex: string) => void;
+    onRestoreData: (appState: AppState) => void;
 }
 
 export type Props = StateProps & DispatchProps;
 
 export interface State {
-    clipboardText: string;
-    secretText: string;
-    canRestore: boolean;
+    backupPassword: string;
+    backupData: HexString;
+    backupInfo: string;
+    appState: AppState | undefined;
 }
 
 export class Restore extends React.PureComponent<Props, State> {
     public state = {
-        clipboardText: '',
-        secretText: '',
-        canRestore: false,
+        backupPassword: '',
+        backupData: '' as HexString,
```

Comment:
That would not work, because `backupData` is a `HexString`, which is a `BrandedString` and that always requires a typecast when creating from the string literal type.

Reply:
We could make `HexString` as a `FlavoredString` which has the definition of `string & { __tag__?: T}`, that is with the `__tag__` being optional, but then it also means everywhere you expect a `FlavoredString` you can pass in a regular string as well.

I chose to make it `BrandedString` because it gives you stronger type checks and that's the price you have to pay for it. I don't mind this, but if you don't like it we can always make a constructor function `makeHexString: (s: string) => HexString`

Reply:
it's not a big deal :-)
But it's still a good practice to be explicit about default state type

Reply:
Done